zapai/sink.go

```diff
@@ -0,0 +1,159 @@
+	q := u.Query()
+	gracePeriod, err := time.ParseDuration(q.Get(paramGracePeriod))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to parse duration parameter")
+	}
+	interval, err := time.ParseDuration(q.Get(paramMaxBatchInterval))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to parse duration parameter")
+	}
+	size, err := strconv.Atoi(q.Get(paramMaxBatchSize))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to convert size parameter to int")
+	}
+	endpoint, err := url.QueryUnescape(q.Get(paramEndpointURL))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to unescape url parameter")
+	}
+	return &SinkConfig{
+		GracePeriod: gracePeriod,
+		TelemetryConfiguration: appinsights.TelemetryConfiguration{
+			InstrumentationKey: u.User.Username(),
+			EndpointUrl:        endpoint,
+			MaxBatchInterval:   interval,
+			MaxBatchSize:       size,
+		},
+	}, nil
+}
+
+// toURI generates a URL from a SinkConfig.
+func toURI(sc *SinkConfig) *url.URL {
+	u := &url.URL{
+		Scheme: SinkScheme,
+		User:   url.User(sc.InstrumentationKey),
+	}
+	q := u.Query()
+	q.Add(paramEndpointURL, url.QueryEscape(sc.EndpointUrl))
+	q.Add(paramGracePeriod, url.QueryEscape(sc.GracePeriod.String()))
+	q.Add(paramMaxBatchInterval, url.QueryEscape(sc.MaxBatchInterval.String()))
+	q.Add(paramMaxBatchSize, url.QueryEscape(strconv.Itoa(sc.MaxBatchSize)))
+	u.RawQuery = q.Encode()
+	return u
+}
+
+// URI builds an appinsights Sink URI string suitable for passing to zap.Open.
+func (sc *SinkConfig) URI() string {
+	return toURI(sc).String()
+}
+
+// telemetryTracker is a client interface that the appinsights.TelemetryClient implements implicitly.
+// It is defined here to restrict the available methods in the Sink from that TelemetryClient, and for testing.
+type telemetryTracker interface {
+	Track(appinsights.Telemetry)
+	Channel() appinsights.TelemetryChannel
+}
+
+var _ zap.Sink = (*Sink)(nil)
+
+// Sink implements zap.Sink for appinsights.
+// Sink is not inherently safe for concurrent use - the zap constructors will wrap it for concurrency.
+//
+// To conform to the zap.Sink interface, Sink implements Write([]byte), where it expects that the passed []byte is
+// an encoded gob that can be decoded in to a valid appinsights.TraceTelemetry. Passing any other []byte input is
+// an error.
+type Sink struct {
+	*SinkConfig
+	cli telemetryTracker
+	dec traceDecoder
+}
+
+// newSink constructs a Sink from the passed SinkConfig
+func newSink(cfg *SinkConfig) *Sink {
+	return &Sink{
+		SinkConfig: cfg,
+		cli:        appinsights.NewTelemetryClientFromConfig(&cfg.TelemetryConfiguration),
+		dec:        newTraceDecoder(),
+	}
+}
+
+// Write accepts a gob []byte that must be Decodable to an appinsights.TraceTelemetry{}, which is then sent
+// to appinsights via the telemetryTracker.
+func (s *Sink) Write(b []byte) (int, error) {
+	t, err := s.dec.decode(b)
+	if err != nil {
+		return -1, errors.Wrap(err, "sink failed to decode trace")
```

Comment:
`-1` looks like an in-band error, is it needed here?

Reply:
0 works just as well, updated.